_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+
+There are two ways to achieve this. The sub chapter 19.2 in <a id="mmds-ref" href="#programming-in-scala">[1]</a> is useful in what follows.  
+
+
+The first is by adding <strong>private</strong> before the parameter list of the class, and then defining an <strong>apply</strong> method on the companion object.  
+
+```scala
+class Reservation[SeatType <: Seat] private (
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+While one does not need to name the method <strong>apply</strong>, it must be on the companion object; otherwise one does not have access to the constructor which is defined as private. Recall that the companion object has the same access rights as the class.
+
+```scala
+object Reservation {
+  def apply[SeatType <:Seat]( userTicket: Ticket[SeatType], guestTicket: Ticket[SeatType]): Option[Reservation[SeatType]] = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => Some(new Reservation(userTicket, guestTicket))
+    case (VipTribune(_), VipTribune(_)) => Some(new Reservation(userTicket, guestTicket))
+    case (_: Standing.type , _: Standing.type ) => Some(new Reservation(userTicket, guestTicket))
+    case _ => None
+  }
+}
+```
+
+Notice that trying to extend Reservation on another file, possibly bypassing our <strong>apply</strong> method, will not work.
+
+As an alternative, one can hide the case class altogether, i.e. not only the constructor but also the name (so that we cannot use it as a type). To that end one defines a trait which exposes the implementation that we want.
+
+
+```scala
+sealed trait Reservation[SeatType <: Seat] {
+  val userTicket: Ticket[SeatType]
+  val guestTicket: Ticket[SeatType]
+}
+```
+
+and on the companion object we have the actual implementation:
+
+```scala
+object Reservation {
+
+  private case class ReservationImpl[SeatType <: Seat](
+    userTicket: Ticket[SeatType],
+    guestTicket: Ticket[SeatType]
+  ) extends Reservation[SeatType]
+
+  def apply[SeatType <: Seat](userTicket: Ticket[SeatType], guestTicket: Ticket[SeatType]): Option[Reservation[SeatType]] = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (VipTribune(_), VipTribune(_)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (_: Standing.type , _: Standing.type ) => Some(ReservationImpl(userTicket, guestTicket))
+    case _ => None
+  }
+}
+```
+
+Notice the relevance of declaring the trait sealed. It prevents the user from extending the trait by another class on some other file, which would allow him to bypass the factory method <strong>apply</strong> above. Equally important is to have ReservationImpl as private so that it cannot be extended on client code. In fact, I guess we could even ignore ReservationImpl altogether and instantiate an anonymous class instead.
```

Comment:
good point here. One oter thing you may consider (not specially here but in some places else) is to make your case classes final to avoid any kind of external subtyping.

Reply:
didn't remember this.  Thanks.